pkg/sql/ttl/ttljob/ttljob.go

```diff
@@ -63,12 +66,104 @@ type rowLevelTTLResumer struct {
 	st  *cluster.Settings
 }
 
+// RowLevelTTLMetrics are the row-level TTL job metrics.
+type RowLevelTTLMetrics struct {
+	RangeTotalDuration *metric.Histogram
+	SelectDuration     *metric.Histogram
+	DeleteDuration     *metric.Histogram
+	RowSelections      *metric.Counter
+	RowDeletions       *metric.Counter
+	NumActiveRanges    *metric.Gauge
+}
+
+// MetricStruct implements the metric.Struct interface.
+func (RowLevelTTLMetrics) MetricStruct() {}
+
+func makeRowLevelTTLMetrics(
+	histogramWindowInterval time.Duration, name string, addMetricStructToRegistry func(interface{}),
+) metric.Struct {
+	sigFigs := 2
+	labels := []*metric.LabelPair{
+		{
+			Name:  proto.String("relation"),
+			Value: proto.String(name),
+		},
+	}
+	ret := RowLevelTTLMetrics{
+		RangeTotalDuration: metric.NewHistogram(
+			metric.Metadata{
+				Name:        "jobs.row_level_ttl.range_total_duration",
+				Measurement: "seconds",
+				Unit:        metric.Unit_NANOSECONDS,
+				MetricType:  io_prometheus_client.MetricType_HISTOGRAM,
+				Labels:      labels,
+			},
+			histogramWindowInterval,
+			time.Hour.Nanoseconds(),
+			sigFigs,
+		),
+		SelectDuration: metric.NewHistogram(
+			metric.Metadata{
+				Name:        "jobs.row_level_ttl.select_duration",
+				Measurement: "nanoseconds",
+				Unit:        metric.Unit_NANOSECONDS,
+				MetricType:  io_prometheus_client.MetricType_HISTOGRAM,
+				Labels:      labels,
+			},
+			histogramWindowInterval,
+			time.Minute.Nanoseconds(),
+			sigFigs,
+		),
+		DeleteDuration: metric.NewHistogram(
+			metric.Metadata{
+				Name:        "jobs.row_level_ttl.delete_duration",
+				Measurement: "nanoseconds",
+				Unit:        metric.Unit_NANOSECONDS,
+				MetricType:  io_prometheus_client.MetricType_HISTOGRAM,
+				Labels:      labels,
+			},
+			histogramWindowInterval,
+			time.Minute.Nanoseconds(),
+			sigFigs,
+		),
+		RowSelections: metric.NewCounter(
+			metric.Metadata{
+				Name:        "jobs.row_level_ttl.rows_selected",
+				Measurement: "num_rows",
+				Unit:        metric.Unit_COUNT,
+				MetricType:  io_prometheus_client.MetricType_COUNTER,
+				Labels:      labels,
+			},
+		),
+		RowDeletions: metric.NewCounter(
+			metric.Metadata{
+				Name:        "jobs.row_level_ttl.rows_deleted",
+				Measurement: "num_rows",
+				Unit:        metric.Unit_COUNT,
+				MetricType:  io_prometheus_client.MetricType_COUNTER,
+				Labels:      labels,
+			},
+		),
+		NumActiveRanges: metric.NewGauge(
+			metric.Metadata{
+				Name:        "jobs.row_level_ttl.num_active_ranges",
+				Measurement: "num_active_workers",
+				Unit:        metric.Unit_COUNT,
+				Labels:      labels,
+			},
+		),
+	}
+	addMetricStructToRegistry(ret)
```

Comment:
So, previously, when I attempted to do this kind of dynamic registration, scoped by name (relationship) in this case, I got some well reasoned pushback from @andreimatei .  I wonder if he has thoughts on this.
Also, I don't know that registry supports adding metrics struct containing maps -- I see it supports array of metric structs, but definitely not maps.

I suspect that perhaps the thing to do here might be to use Agg* metrics, like we do in changefeedccl/metrics?

Reply:
I don't see a way of doing this using aggmetric as we don't know the table names upfront. Also note we only keep using the map to reference the metrics - the MakeRowLevelTTLMetrics function adds it to the server registry. The map is simply used to reference the metric struct internally.

Reply:
ah i misread aggmetric - i'll look later

Reply:
The aggmetric labels don't seem to show up in /_status/vars - - do you know where they actually show up?